Tooltips: Make tooltips non interactive by default

torkelo opened the pull request:
Our current tooltips can often create a really bad user experience because they do not automatically hide when you hover out from the trigger. If you want to interact with something behind the tooltip you can't. 

This has been raised in an issue can't find it though. 

Video describing/showing the biggest problem 
https://user-images.githubusercontent.com/10999/152971612-ce4e904d-f247-4f82-a4c2-72bb4d3faa2d.mp4

* Updates our Tooltip to not use PopoverController (think we should deprecate this) 
* Uses usePopperTooltip as it has all the features we need (simple abstractions ontop of usePopper) 
* Show tooltip on focus not just hover? (good idea? option?) 
* Adds an interactive property that is default false that controls if a tooltip can be hovered onto (and interacted with, ex click a link)

There is no critical breaking change (should not be), but a breaking behavior here. So existing Tooltip uses that have links inside need to add the property `interactive` to get the same behavior. 

# Release notice breaking change

The Tooltip component provided by `@grafana/ui` is no longer automatically interactive (that is you can hover onto it and click a link or select text). It will from now on by default close automatically when you mouse out from the trigger element. To make tooltips behave like before set the new `interactive` property to true.

torkelo committed:
Tooltips: Make tooltips non interactive by default

dprokop commented on packages/grafana-ui/src/components/Tooltip/Tooltip.tsx:
```diff
@@ -1,61 +1,189 @@
+    delayHide: interactive ? 100 : 0,
+    delayShow: 50,
+    // Focus is new, old tooltip did not show on focus, we could make this a parameter/option?
+    trigger: ['hover', 'focus'],
```
From the a11y POV I think we should keep the focus to open the tooltips automatically. Can't think of any situation when provided tooltip would not be shown in keyboard navigation scenario.

So IMO the behavior presented by this PR is desired one.

dprokop commented on public/app/features/dashboard/dashgrid/PanelHeader/PanelHeaderCorner.tsx:
```diff
@@ -86,7 +86,7 @@ export class PanelHeaderCorner extends Component<Props> {
 
     return (
-      <Tooltip content={content} placement="top-start" theme={theme}>
+      <Tooltip content={content} placement="top-start" theme={theme} interactive>
```
Interactive works as expected 👍

dprokop reviewed (approved):
LGTM. Think this is a nice improvement, was frustrating to me as well. Also - I like the subtle design update to the tooltips, they now look more refined.

torkelo committed:
More updates and cleanup

leeoniya commented:
cc @nmarrs is doing a lot of tooltip work currently. also possibly related: https://github.com/grafana/grafana/issues/45032

aocenas commented:
Copying this from slack:

personally would be for keeping the interactivity. Especially in the case @torkelo shown in the video there is lots of info in those tooltips and I hate when I have elements in the page with docs for example from which I cannot copy paste stuff.

Like if it is just to give a name to an icon button, that is probably no issue, but we tend to put lots of docs into those tooltips lots of times it seems.

For this particular issue, I would rather have those info buttons to open the tooltip on click instead on hover for example which I think would solve this and be probably more UX friendly (like if you are poweruser you don't need those at all so it does not make sense to even flash them when moving your mouse around)

torkelo committed:
Update

torkelo committed:
Make time range picker tooltip interactive

torkelo committed:
Fix issue

axelavargas reviewed (commented):
Hey Torkel 👋🏾 , here are some thoughts after taking a look:

- Making non-interactive by default: I think adding this change will potentially break existing  use cases that we can miss during this PR, for example when we use tooltip JSON ([today in Grafana play](https://play.grafana.org/d/0000000121/table-json-tooltip-closes-when-extended-34553?orgId=1))

And this is how will look with this change: 
![tooltipJson](https://user-images.githubusercontent.com/239999/153024121-a95a72e6-2baf-4f81-878b-bf17c96a717b.gif)

Of course, this just opens a different discussion, I think this problem is the consequence of using Tooltips and Popovers as the same thing, **maybe this is a good opportunity of making a clear difference?**

Tooltip: When we want to show a short text, it will be visible when hovering or focusing over a word or icon, should not receive focus.

Popover: Show a longer text, or when you want to have a link to an external web page. It is shown when clicking over a word or icon. The window remains visible until they click again outside of the popover.

An idea 💡 :
We could create 2 new components (TooltipV2 and Popover) and use them correctly? 


> Show tooltip on focus not just hover? (good idea? option?) 

Based on [a11y guidelines is recommended we do that](https://dequeuniversity.com/library/aria/tooltip)

torkelo committed:
Trying to make Receivers test faster

torkelo committed:
Make JSONCell tooltip interactive

torkelo commented:
@axelavargas I have fixed the JSONView tooltip, good catch, added the new property ìnteractive` to this tooltip so you can hover onto it like before. 

Yea, think it might be a good idea to create a separate Popover component (we have one but it's not great) for larger Tooltips. But can explore that in another PR
